### ISSUE TITLE
About train.py code

### ISSUE THREAD

**Issue by 2000Zzw** [nie-lang/StabStitch#3]
Hello, I want to train my own datasets, it is greatly appreciated that you would give the training codes.

**Comment by nie-lang** [comment]
Hi, thanks for your interest. The related training code of StabStitch has not been organized. Actually, I'm working on the extension version of StabStitch with better alignment, fewer distortions, and higher stability. We plan to release the complete code of the extension (including training and testing codes) once this work is done, whether the extended paper is accepted or not.

**Comment by 2000Zzw** [comment]
Thanks for your Reply.

**Comment by 2000Zzw** [comment]
> Hi, thanks for your interest. The related training code of StabStitch has not been organized. Actually, I'm working on the extension version of StabStitch with better alignment, fewer distortions, and higher stability. We plan to release the complete code of the extension (including training and testing codes) once this work is done, whether the extended paper is accepted or not.

I have some questions on the code.
1、When I run the 'test_online.py' code, I found the output of the '_transform' in the 'torch_tps_transform.py' is a full zero matrix. So I normalize the 'x_s' and 'y_s'. I'm not sure if it is the correct operation.

![image](https://github.com/user-attachments/assets/39f93dfb-6a42-480e-9667-29a11a730e39)


2、I found it takes a lot of time to generate one stitching video(I ran it for a night but without a stitching video has been generate in the 'result'. ).

**Comment by nie-lang** [comment]
> > Hi, thanks for your interest. The related training code of StabStitch has not been organized. Actually, I'm working on the extension version of StabStitch with better alignment, fewer distortions, and higher stability. We plan to release the complete code of the extension (including training and testing codes) once this work is done, whether the extended paper is accepted or not.
> 
> I have some questions on the code. 1、When I run the 'test_online.py' code, I found the output of the '_transform' in the 'torch_tps_transform.py' is a full zero matrix. So I normalize the 'x_s' and 'y_s'. I'm not sure if it is the correct operation.
> 
> ![image](https://private-user-images.githubusercontent.com/169328807/363386088-39f93dfb-6a42-480e-9667-29a11a730e39.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjUwNzczMjYsIm5iZiI6MTcyNTA3NzAyNiwicGF0aCI6Ii8xNjkzMjg4MDcvMzYzMzg2MDg4LTM5ZjkzZGZiLTZhNDItNDgwZS05NjY3LTI5YTExYTczMGUzOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODMxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgzMVQwNDAzNDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNzc0OWFhNDlmNzllOWNiZWIxNzk2YjQzODU5MTJjYzZkMDBmZDAyNThjYzhiMWMyNjFiOWYyNjAzYjhlMjQ4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.lKeRTUuvzgxY1E-9zkqhu0P7kJQU0DJPFqjDTmNB2m0)
> 
> 2、I found it takes a lot of time to generate one stitching video(I ran it for a night but without a stitching video has been generate in the 'result'. ).

Please make sure the input video frames and models are correctly loaded. I tested it on a 4090 GPU and it could run in real-time.

**Comment by 2000Zzw** [comment]
Thanks! It could work!

**Comment by liujiaocv** [comment]
I am extremely grateful for your outstanding work. May I inquire about the specific date for the release of the training code?

**Comment by nie-lang** [comment]
> I am extremely grateful for your outstanding work. May I inquire about the specific date for the release of the training code?

If all things go well, we will release the complete code for the extended version in October.

**Comment by yfclark** [comment]
@nie-lang 
The open-source project is excellent, and I am currently testing it. There are some areas where the code could be improved:

(1) Directly constructing tensors on CUDA can be slightly faster than first constructing them on the CPU and then moving them to the GPU.

(2) The `test_online` logic still has some issues. The spatial warp and temporal warp should be processed in a rolling manner, rather than processing the entire video in one batch and then passing it to the next step.
Lastly, I look forward to seeing the code for the training part at an earlier stage.

**Comment by nie-lang** [comment]
@yfclark 


> @nie-lang The open-source project is excellent, and I am currently testing it. There are some areas where the code could be improved:
> 
> (1) Directly constructing tensors on CUDA can be slightly faster than first constructing them on the CPU and then moving them to the GPU.
> 
> (2) The `test_online` logic still has some issues. The spatial warp and temporal warp should be processed in a rolling manner, rather than processing the entire video in one batch and then passing it to the next step. Lastly, I look forward to seeing the code for the training part at an earlier stage.

First of all, thanks for your suggestions. As for the logic of "test_online", we first calculate all the spatial/temporal warps in the whole videos to determine the final size of the stitched video. Then we warp all frames with this size. Actually, we can process the videos in one batch and then pass it to the next step as you said. But this way, we should predefined the resolution of the stitched video. Typically, to ensure all content not lost， the predefined resolution would be pretty large, which would produce extensive invalid regions (black regions) and decrease the warping speed.

Finally, we will release the training code of the extended version (StabStitch++) within this October.

**Comment by nie-lang** [comment]
The complete code of [StabStitch++](https://github.com/nie-lang/StabStitch2) (an extension of StabStitch) is released, including the codes for training, inference, and multi-video stitching.
